File: 98/p98.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <inttypes.h>

typedef struct{char set[26], word[15];} Word;

//1786 words, 14 letters longest
char isSquare(uint64_t n){
	int r = sqrt(n);
	return r*r == n;
}

void makeSet(const char *w, char *s){
	while(*w){
		++s[*w - 'A'];
		++w;
	}
}

int compareBySet(const Word *a, const Word *b){
	return memcmp(a->set, b->set, sizeof(a->set));
}

int compareByLen(const Word *a, const Word *b){
	int d = strlen(b->word) - strlen(a->word);
	if(d){
		return d;
	}
	return compareBySet(a, b);
}

int processCombination(Word a, Word b){
	
}

int processCombinations(Word *words, int a, int b){
	int max = 0, cur;
	for(int i = a; i < b - 1; ++i){
		for(int j = i + 1; j < b; ++j){
			cur = processCombination(words[i], words[j]);
			if(cur > max){
				max = cur;
			}
		}
	}
	return max;
}

int main(void){
	FILE *file = fopen("words.txt", "r");
	if(!file){
		puts("could not open file");
		return 1;
	}
	Word words[1786];
	for(int i = 0; i < 1786; ++i){
		fscanf(file, "%26s", words[i].word);
		makeSet(words[i].word, words[i].set);
	}
	fclose(file);
	qsort(words, 1786, sizeof(Word), (__compar_fn_t)compareBySet);
	Word anagrams[1786];
	int l = 0;
	for(int i = 0; i < 1786; ++i){
		if(compareBySet(words + i, words + i + 1)){
			continue;
		}
		while(!compareBySet(words + i, words + i + 1)){
			anagrams[l++] = words[i++];
		}
		anagrams[l++] = words[i];
	}
	qsort(anagrams, l, sizeof(Word), (__compar_fn_type)compareByLen);
	int max = 0;
	for(int i = 0, j = 1; i < l; i = j++){
		while(!compareBySet(words + i, words + j)){
			++j;
		}
		if((max = processCombinations(words, i, j))){
			break;
		}
	}
	printf("%i\n", max);
}

File: 98/p98.py
from itertools import groupby, combinations, permutations
from math import sqrt

def letters(word):
	ret = [0]*26
	for letter in word:
		ret[ord(letter) - ord('A')] += 1
	return ret

def fst(t):
	return t[0]

def snd(t):
	return t[1]

def isSquare(n):
	r = int(sqrt(n))
	return r*r == n

def ltoi(digits):
	n = 0
	for digit in digits:
		n *= 10
		n += digit
		if n == 0:
			return 0
	return n

with open("words.txt", "r") as f:
	words = f.read().splitlines()
	words = [(letters(word), word) for word in words]
	words.sort(key = fst)
	aFamilies = groupby(words, key = fst)
	aFamilies = ([word for mask, word in family] for key, family in aFamilies)
	anagrams = [(sorted(set(family[0])), family) for family in aFamilies if len(family) > 1 and len(family[0]) >= 4]
	anagrams.sort(key = lambda family: -len(snd(family)[0]))
	maxSquare = 9216
	maxLen = 4
	for letters, family in anagrams:
		if len(family[0]) < maxLen:
			break
		for a, b in combinations(family, 2):
			print(a, b)
			for digits in permutations(range(10)[::-1], len(letters)):
				ltod = dict(zip(letters, digits))
				inta = ltoi(map(ltod.get, a))
				if inta == 0:
					continue
				if not isSquare(inta):
					continue
				intb = ltoi(map(ltod.get, b))
				if intb == 0:
					continue
				if not isSquare(intb):
					continue
				print("a and b square")
				print(digits)
				print(inta)
				print(intb)
				inta = max(inta, intb)
				if inta > maxSquare:
					maxSquare = inta
					maxLen = len(family[0])
	print(maxSquare)
